opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -5,139 +5,167 @@ namespace Opm {
-                      const bool terminal_output,
-                      const int current_timeIdx,
-                      const std::vector<int>& pvt_region_idx)
-       : wells_active_(wells_arg!=nullptr)
-       , wells_(wells_arg)
-       , wells_ecl_(wells_ecl)
-       , number_of_wells_(wells_arg ? (wells_arg->number_of_wells) : 0)
-       , number_of_phases_(wells_arg ? (wells_arg->number_of_phases) : 0) // TODO: not sure if it is proper for this way
-       , param_(param)
-       , well_container_(createWellContainer(wells_arg, wells_ecl, param.use_multisegment_well_, current_timeIdx, param) )
-       , well_collection_(well_collection)
-       , terminal_output_(terminal_output)
-       , has_solvent_(GET_PROP_VALUE(TypeTag, EnableSolvent))
-       , has_polymer_(GET_PROP_VALUE(TypeTag, EnablePolymer))
-       , current_timeIdx_(current_timeIdx)
-       , rate_converter_(rate_converter)
-       , pvt_region_idx_(pvt_region_idx)
+                      const bool terminal_output)
+        : ebosSimulator_(ebosSimulator)
+        , param_(param)
+        , terminal_output_(terminal_output)
+        , has_solvent_(GET_PROP_VALUE(TypeTag, EnableSolvent))
+        , has_polymer_(GET_PROP_VALUE(TypeTag, EnablePolymer))
     {
-    }
+        const auto& eclState = ebosSimulator_.gridManager().eclState();
+        phase_usage_ = phaseUsageFromDeck(eclState);
+
+        active_.resize(phase_usage_.MaxNumPhases, false);
+        for (int p = 0; p < phase_usage_.MaxNumPhases; ++p) {
+            active_[ p ] = phase_usage_.phase_used[ p ] != 0;
+        }
+
+        const auto& gridView = ebosSimulator_.gridView();
 
+        // calculate the number of elements of the compressed sequential grid. this needs
+        // to be done in two steps because the dune communicator expects a reference as
+        // argument for sum()
+        number_of_cells_ = gridView.size(/*codim=*/0);
+        global_nc_ = gridView.comm().sum(number_of_cells_);
+        gravity_ = ebosSimulator_.problem().gravity()[2];
 
+        extractLegacyCellPvtRegionIndex_();
+        extractLegacyDepth_();
+    }
 
 
 
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::
-    init(const PhaseUsage phase_usage_arg,
-         const std::vector<bool>& active_arg,
-         const double gravity_arg,
-         const std::vector<double>& depth_arg,
-         long int global_nc,
-         const Grid& grid)
+    beginReportStep(const int timeStepIdx)
     {
-        // has to be set always for the convergence check!
-        global_nc_   = global_nc;
+        const Grid& grid = ebosSimulator_.gridManager().grid();
+        const auto& defunct_well_names = ebosSimulator_.gridManager().defunctWellNames();
+        const auto& eclState = ebosSimulator_.gridManager().eclState();
+        wells_ecl_ = eclState.getSchedule().getWells(timeStepIdx);
+
+        // Create wells and well state.
+        wells_manager_.reset( new WellsManager (eclState,
```

Comment:
`WellsManager` is created each report step, it does not need to be a member. `WellCollection` can be a smaller member to keep.

Reply:
after this PR, the well model object is persistent for the whole simulation. this is done in order to be able to fold the `WellState`/`struct Wells`/`WellManager`/`WellCollection`/`RateConverter`/`VFP` logic into the `WellModel` class so that all of this stuff can be refactored/removed without touching outside code. this is the primary semantic change of this PR ;)

Reply:
Sure. It is just nitpickking for discussion. I see now `Wells struct` is also from `wells_manager_`.

Reply:
> Sure. It is just nitpickking for discussion.

That's how I interpreted you. Thanks for having a thorough look, BTW; sometimes it is hard to see such issues in changes if they are too familiar.